### DISCUSSION TITLE
Why draw_messages() in src/streamlit_app.py expects only human and ai types?

### DISCUSSION THREAD

**Discussion by 7sunstudio** [JoshuaC215/agent-service-toolkit#39] [Q&A] [answered]
In schema.py ChatMessage class, type could be human, ai or tool. But draw_messages() will throw error on tool type. Did I miss something somewhere? I read through the code multiple times but couldn't figure out why.

**Comment by 7sunstudio** [comment]
After more debug, I found out the magic happens on line 198 in streamlit_app.py
`if is_new:
st.session_state.messages.append(msg)`

If I modified the condition to `if is_new and msg.content != ""`, some messages have type "tool" and others have type "ai".

I still don't understand why appending content to session_state will impact message type though.

**Comment by JoshuaC215** [comment, marked as answer]
The expected pattern is that a ToolMessage will follow an AIMessage that contains `tool_calls`, so the ToolMessage is handled with an inner iteration rather than in the main loop - see [here](https://github.com/JoshuaC215/agent-service-toolkit/blob/98def1ad8aede56aea1cd3cb271316edfb43ec1d/src/streamlit_app.py#L218-L238).

However I recently saw some unexpected errors in the current version of the app when the research assistant requests multiple parallel tool calls. This used to work but seems to be broken now. I haven't had a chance to debug but it does suggest that this code isn't perfect currently :)

**Comment by 7sunstudio** [comment]
Thanks for the quick response, @JoshuaC215 .

Yes, I also saw multiple tool calls are triggered in some of the request. I reckon it might be caused by the change on Langraph.
